### ISSUE TITLE
FR: Add a user hyphenation directory additional to the provided ones.

### ISSUE THREAD

**Issue by zwim** [koreader/koreader#7722]
**Some words especially some names get hyphenated wrong**
Normally in German the hyphenation patterns are quite good. But sometimes words get hyphenated wrong (especially unusual names).

**Describe the solution you'd like**
I would like to add a user hyphenation dictionary. Words in that dict have a higher priority than the provided patterns. So if there is a word in that dict the provided patterns are skipped.

For each language there is a different user dict.

Please see my proposed solution (video) below. I have it working now, but would like to do some more tests (esp. with corner cases).

I would appreciate your comments about the UI. 
I will open a PR as soon as my tests are finished.

**Describe alternatives you've considered**
Add every and each word to the provided dictionary. This would be quite time-consuming (in German there is a project doing that, our German patterns come from there) and error-prone. You would have to live with the wrong hyphenations until the next update.

**Demonstration**
This is just a demonstration about the functionality. The chosen word "Selbstbiographie" may be hypenated "Selb-st-bio-gra-phie", by our provided pattern. But I can play with :)

https://user-images.githubusercontent.com/36999612/119017675-88230d00-b99b-11eb-8578-c97c9a6423fa.mp4

**Comment by poire-z** [comment]
Dunno, feels overkill to the not-german I am.
And I would add words that I see only once and never meet again. And I would keep these fix to words for myself, and forget about contributing to the main hyph dict :/

The good is that there' no UI noise if you keep that checkbox unchecked.
But I fear the level of interaction readertypography <> dictquicklookup <> crengine (and obviously for me, how light is this support in crengine).

**Comment by zwim** [comment]
@poire-z: The patch really managable: 
in readertypograph 3 lines, 
in readerui +1 registered module, 
in credocument 2 wrapper functions (8 lines), 
in quickdictlookup 1 line
and a new module readeruserhyph with about 200 lines mostly menu to add entry, format and checks.

in cre.cpp 1 wrapper function

in lvdocview.cpp and lvviewprops.cpp +1 property
in lvhyphman.cpp one new class 160 lines of code. (mainly one function call).

From my view this is very light.

**Comment by poire-z** [comment]
Some other idea instead of hacking DickQuickLookup - as most often, you'll be doing real dict lookups and you don't need to see that button:
add it to ReaderHighlight's dialog. You'll need to very-long-press on a word to get it for single word selection (> 3/4 seconds) - and may be (haven't checked) at the time you build this dialog's buttons, you can see if it's a single word or not and add or not this other button.

**Comment by zwim** [comment]
The idea sounds good. But if I try this in the emulator, always the dictlookup dialog pops up. 

Only if I press for >4sec. and move the cursor while (at the end of) pressing the highlight dialog shows up. Is this behavior wanted?

**Comment by roshavagarga** [comment]
Hmm, I can see how this could be a useful addition, but I'd note it would need to be documented well enough for complete beginners to make a dict on their own. I guess it could also be fairly useful for additional dictionaries we don't want to include hyphen patterns for? For instance, say there's a German dictionary that's strictly useful for people reading medical or judicial things - in those use-cases we can redirect users to this option, thus not having to making existing dicts bigger or customizing them from multiple sources. Of course, I might be misunderstanding everything I've read so far, in which case chalk up anything dumb I've said to my typical lack of dev knowledge 👍

**Comment by zwim** [comment]
> Edit: I don't remember if we ever cut down on those mixed Russian+English dicts, but can't we use this feature to do that somehow?

If we have the wordlist of the dictionaries, it should be doable. I have to think about it.

**Comment by poire-z** [comment]
> But if I try this in the emulator, always the dictlookup dialog pops up.

Same for me. It used to work 2 months ago, so a possible regression from these last 2 months, I'll look at it.

**Comment by zwim** [comment]
Thanks, no hurry.

Am 20. Mai 2021 21:11:51 MESZ schrieb poire-z ***@***.***>:
>> But if I try this in the emulator, always the dictlookup dialog pops
>up.
>
>Same for me. It used to work 2 months ago, so a possible regression
>from these last 2 months, I'll look at it.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/koreader/koreader/issues/7722#issuecomment-845404958

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by poire-z** [comment]
> Edit: I don't remember if we ever cut down on those mixed Russian+English dicts, but can't we use this feature to do that somehow?

No, we haven't, I still have that in my to-look-at list.

> If we have the wordlist of the dictionaries, it should be doable. I have to think about it.

I have yet no idea how you implemented this feature, but for any merging of EN+RU, the most obvious  idea was to merge the already made individual EN and RU pattern files, and not have to worry making new patterns out of new wordlists.

**Comment by poire-z** [comment]
@NiLuJe : any science backing this is the solution to the above issue ? :)
```lua
--- a/frontend/apps/reader/modules/readerhighlight.lua
+++ b/frontend/apps/reader/modules/readerhighlight.lua
@@ -1156,7 +1156,7 @@ end
 function ReaderHighlight:onHoldRelease()
     local long_final_hold = false
     if self.hold_last_tv then
-        local hold_duration = UIManager:getTime() - self.hold_last_tv
+        local hold_duration = TimeVal.now() - self.hold_last_tv
         if hold_duration > TimeVal:new{ sec = 3, usec = 0 } then
             -- We stayed 3 seconds before release without updating selection
             long_final_hold = true
```

Adding `logger.warn(TimeVal.now() - UIManager:getTime())` just before, and holding for ~4 seconds shows:
```
05/20/21-21:26:05 DEBUG CreDocument: goto page 6 flow 0
05/20/21-21:26:06 DEBUG refresh on physical rectangle 51 272 63 17
05/20/21-21:26:10 DEBUG input event => type: 3 (EV_ABS), code: 57 (ABS_MT_TRACKING_ID), value: -1, time: 4609284.858589
05/20/21-21:26:10 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 4609284.858589
05/20/21-21:26:10 DEBUG in hold state...
05/20/21-21:26:10 DEBUG hold_release detected in slot 0
05/20/21-21:26:10 DEBUG adjusted ges: hold_release
05/20/21-21:26:10 WARN  {
    ["sec"] = 4,
    ["usec"] = 621651,
}
```

**Comment by zwim** [comment]
TimmeVal:now() : NOT . ?

**Comment by NiLuJe** [comment]
There's no longer UI ticks in between a single batch of input events uninterrupted by a scheduled task since the batched input, so, yeah, that probably make sense ;).

**Comment by poire-z** [comment]
Oh yes, it is still working on my Kobo :/ So, the issue was just on the emulator?

**Comment by NiLuJe** [comment]
Or if you have a very steady fingers ^^.

**Comment by poire-z** [comment]
Implemented by #7746 / #7787.
